### ISSUE TITLE
yolov8 is working very slowly

### ISSUE THREAD

**Issue by ArpitaG10** [ultralytics/ultralytics#3097]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

yolov8 is working very slowly, i just installed yolov8 and only 8 epoch completed in 1 plus hour, please suggest yolov8 version which is successfully running

### Additional

I already worked on yolov8 before it will take 3-4 hours for 180 images but not it is taking infinte amout to complete 180 images. Please let me know if i can contact you. I need to submit the project soon in my office.

**Comment by glenn-jocher** [comment]
@ArpitaG10 these methods may help you improve inference speeds:

1. **Model Optimization**: Use lighter versions of the model if speed is a priority. These models often sacrifice a bit of accuracy for speed.

2. **Batch Processing**: Use batch processing if you have multiple images to process, as processing many images at once can speed up inference.

3. **Hardware Acceleration**: Utilize a powerful GPU, if possible. You can also try other hardware accelerators like Google's TPUs or edge devices like NVIDIA's Jetson series.

4. **Image Resolution**: Decrease the resolution of your images. A smaller image requires less computation, but be mindful that this could also affect model accuracy.

5. **Quantization**: Quantization reduces the numerical precision of the model's parameters, thus speeding up inference. Common methods include float16 (also known as half-precision) and int8 quantization.

6. **Pruning**: Pruning involves reducing the size of the neural network by removing less-important neurons, thus speeding up inference.

7. **Use ONNX, TensorRT, or TFLite**: Converting your model to ONNX, TensorRT, or TFLite can help optimize for inference speed, especially on specific devices. 

Remember, it's often a tradeoff between speed and accuracy, so choose the best approach that suits your specific requirements.

**Comment by ArpitaG10** [comment]
I am a newbie in this field, please suggest how i can use lighter version of yolov8. And one more thing is in yolov8 only code is using to train is 
"from ultralytics import yolo
 
model = YOLO('yolov8m.pt')

model.train(data='data_custom.yaml', image=640, epochs=150,batch=8) "

in yaml, i just mention class name and number and path of train and val images.

And from past 1 hour only 10 epoch completed, So,please suggest me how i can increase the speed of training. 

lots of thanks.

**Comment by glenn-jocher** [comment]
@ArpitaG10 hello and welcome to YOLOv8!

For using a lighter version, you can either use the pre-trained checkpoint which was trained on a smaller dataset or you can simply modify the model architecture by modifying the `num_filters` and `num_blocks` in the model configuration file. This way your model will be smaller, lighter and faster at the cost of some accuracy. You can find details on this in the [YOLOv8 Documentation](https://docs.ultralytics.com/yolov8/models/).

Regarding the slow training, training speed can depend on many factors, including the dataset size, batch size, and hardware. Here are a few things to try which may help increase the training speed:

1. Increase the batch size: If you have enough memory to fit in larger batches, try increasing the batch size during training. This is a great way to utilize available hardware and speed up your training.

2. Modify training configurations: Some training configurations may produce faster inference times while sacrificing some accuracy.  You can try modifying some training configurations, such as learning rate and optimizer, for a faster training speed. Also, ensure that your GPU is being utilized fully during training.

3. Use Data Augmentation Techniques: Data augmentation techniques like cropping, flipping, rotating, and scaling can help increase the training data and speed up the training process. You can find several data augmentation techniques in the [YOLOv8](https://docs.ultralytics.com/yolov8/data-augmentation/).

I hope these suggestions help. Don't hesitate to ask if you need further assistance.

**Comment by ArpitaG10** [comment]
Sorry if i am asking silly questions but i really couldnt able to find the "model configuration file". 
After completed the training only runs folder created and inside runs folder i found weights and other matrics but "configuration file " icouldnt able to find. Please guide me.

And thanks for answering every time.

**Comment by glenn-jocher** [comment]
@ArpitaG10 if you are using YOLOv8m (medium) then you can downsize to a smaller model like YOLOv8n (nano):

```
model = YOLO('yolov8n.pt')
```

See our README table for model sizes.

**Comment by glenn-jocher** [comment]
@Ritesh091 the time taken for quantization can depend on numerous factors. The primary reason could be the size of the neural network. Bigger and complex models will take a longer time to quantize. The hardware being used for the process also plays a significant role. High-performance machines and GPUs will be quicker in performing computations required for quantization. 

There are a couple of things you can do to potentially speed up the process:

1. **Use a smaller model**: If your model is too complex, using a smaller, simpler model can speed up the quantization process. YOLOv8 offers different sizes of models, so choosing a smaller one might help.

2. **Optimize the quantization parameters**: You may be using parameters or techniques that are making the quantization process longer. Revisiting and optimizing these parameters might help. Parametrization depends on specific use-cases, so you'll need to experiment with different configurations to find what works best for you.

3. **Hardware used**: As mentioned, the hardware being used can significantly impact the time taken for quantization. Using a more powerful machine and/or GPU can dramatically speed up the process.

Lastly, also please note that the quantization process is inherently complex and computationally intensive. Not all processes can be substantially sped up beyond certain limits. Patience will be needed in many cases. Hope this helps!

**Comment by raedabusanad** [comment]
Hello @glenn-jocher ,
I have downloaded the Yolov8 from Ultralytics website and converted it to ONNX. 
Any guidance on how can i optimise its quantization parameters?

**Comment by glenn-jocher** [comment]
@raedabusanad, quantization parameters optimization generally involves selecting the right balance between performance and accuracy. When you convert a model to ONNX and then to a quantized format like INT8, there are some considerations you should keep in mind:

1. **Calibration Dataset**: Select a representative calibration dataset that captures the variability in your actual dataset. This is crucial as the calibration process will determine which values to scale down during quantization.

2. **Quantization Method**: There are different quantization schemes like dynamic, static (also known as calibration or post-training quantization), and quantization-aware training (QAT). Static quantization generally provides a good balance between performance and accuracy.

3. **Calibration Steps**: Run the calibration using enough steps to ensure the collected statistics are representative of the actual data distribution. Fewer steps might speed up the quantization process, but at the expense of precision and the final accuracy of the model.

4. **Tooling**: Different tools and frameworks (like ONNXRuntime, PyTorch, TensorFlow, etc.) provide their own utilities and parameters for model quantization. Familiarize yourself with the tools specific to ONNX and adjust parameters like quantization algorithms, weight bit-width, activations bit-width, and fusing options.

5. **Precision Requirements**: Depending on your use case, determine if full INT8 quantization is essential. Sometimes a mixed-precision model where some layers remain as FP32 or FP16 can provide a good trade-off between speed and accuracy.

6. **Iterative Approach**: Quantization is often an iterative process. You might need to quantize, test, and then re-adjust your parameters multiple times to find the optimal settings.

Remember, quantization can lead to accuracy degradation. Always validate the performance of the quantized model on a comprehensive test dataset to ensure that it meets your application requirements.

For more detailed guidance, please see our [Quantization](https://docs.ultralytics.com/models/quantization/) documentation. Always be prepared to experiment and potentially compromise in order to find the best setup for your specific needs. Good luck!

**Comment by choudhary4296** [comment]
I had the same problem, 1 epoch was taking 1 hour. This happened when I was using the root user's Torch environment.

The solution I found was:
I configured Torch and Ultralytics standalone without the environment and it worked for me. Training time is reduced, i.e. 1 epoch now takes only 10 minutes. Also, it is now working properly in my user's environment.

Why was this happening through the root?
But it worked for me.

**Comment by glenn-jocher** [comment]
@miss-choudhary hello there! 🌟

Glad to hear you found a solution that significantly reduced your training time! 🚀 The issue you encountered could be related to different environment configurations between the root user and a standard user, potentially affecting Torch's performance due to variations in installed packages, permissions, or hardware access optimizations.

Running models outside a root environment can sometimes yield better performance due to these configuration nuances. Each setup can behave differently based on how dependencies are managed and executed. Your discovery emphasizes the importance of experimenting with different setups to find what works best. 

If you're exploring further optimizations, remember to keep an eye on batch sizes and the model variant you're using, as they can also impact training speed. Happy coding, and feel free to share any more insights or questions you have! 

Keep up the great work. 👍

**Comment by achuntelolan** [comment]
I trained a Yolo v8 n model with a custom data set and converted it into openvino format after that when I tested it with a video for getting compare with Yolo v4 the v8 model took 126sec to complete the whole video but Yolo v4 taking just 90 secs the v4 model image passing size is 480 but v8 is  640 why it's taking more time comparing to yolo v4

**Comment by glenn-jocher** [comment]
@achuntelolan hi there! 😊

The difference in processing time between YOLOv8 and YOLOv4 could likely be due to the larger image size (640 for v8 vs. 480 for v4) you're using. YOLOv8 models, especially with larger input sizes, often have more detailed architectures leading to longer inference times. 

To make YOLOv8 faster, you might consider reducing the input size similar to your YOLOv4 setup or optimizing your model further for performance. 

For example:
```python
model.export('path/to/your/model', imgsz=(480, 480), format='openvino')
```

Give it a try and see if it helps with the speed! 🚀

**Comment by achuntelolan** [comment]
thank you

**Comment by achuntelolan** [comment]
Hii i converted to openvino by changing the size of images but that time I lost my accuracy

**Comment by glenn-jocher** [comment]
@achuntelolan hello there! 👋

Thanks for letting us know about the accuracy drop after converting to OpenVINO with a different image size. This is a known trade-off when optimizing models for speed. However, you can try a few things to mitigate the accuracy loss:

1. **Fine-tuning**: After converting your model, fine-tune it with the smaller image size on your dataset.
2. **Keep Aspect Ratio**: Ensure that when resizing images, you're maintaining the original aspect ratio to prevent distortion that could affect accuracy.

Example for maintaing aspect ratio during resize:
```python
from PIL import Image

image = Image.open('path/to/image.jpg')
image = image.resize((480, 480), Image.ANTIALIAS)
```
Hope this helps! Keep experimenting, and feel free to reach out with any more questions or insights. 😊

**Comment by achuntelolan** [comment]
@glenn-jocher Thank you for your valuable support if you don't mind can you provide your WhatsApp number just text me on WhatsApp if you don't have any issues. my number is +91 9188462190

**Comment by MiessiGomes** [comment]
> Hey there! I'm glad I could help! 🙌 However, for support and questions, let's keep the conversation here or on the GitHub discussions for better tracking. Plus, it helps the whole community learn from each other's questions and insights. Feel free to drop your queries or share your updates, and I or someone from the community will be more than happy to assist! 😊

Is there any other way to increase the training speed without changing this following core specs? 
Core Specs: I'm already using the RTX 3060 12GB, batch = 8, epochs = 200, patience = 100, imgsz = 640, model = yolov8m

For now my model is too small and each epochs are around 11 seconds and the total time was 36 minutes to train 200 epochs, I've tried to use the ONNX format with TensorRT, but the time increased a lot.

My question is, exist another parameter that I can config to increase the training speed ? 
BTW YoloV8 is impressive, congrats !!

**Comment by pderrenger** [comment]
Thank you for your kind words about YOLOv8! To increase training speed without altering your core specs, you might consider enabling mixed precision training by setting `amp=True` in your training configuration. This can significantly speed up training by utilizing both 16-bit and 32-bit floating-point types. Additionally, ensure that your data loading is optimized by setting the `workers` parameter to a higher value, which can help in faster data preprocessing. If you haven't already, you might also want to experiment with the `cache` parameter to store dataset images in memory, reducing disk I/O time. If the issue persists, please ensure you are using the latest versions of all packages.

**Comment by achuntelolan** [comment]
Hi while iam trying to installing the ultrlytics 8.0.128 version for my python3.6 its showing version not found error why can anyone help me for this issue

**Comment by glenn-jocher** [comment]
The Ultralytics package requires Python 3.7 or later. Please upgrade your Python version to 3.7 or higher to install the latest version of Ultralytics.
